### PR TITLE
adding-pipelines

### PR DESCRIPTION
**Pull request by pallavipannu**

### CODE REVIEW

File: external-data/external-data.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            namespace="kubeflow",
+            input_volumes=input_volumes,
+            output_volumes=json.dumps(
+                ["{{workflow.uid}}-featureset@featureset://" + str(featureset)]
+            ),
+        ).after(preprocessing)
+
+        list_featureset = kfp.dsl.ContainerOp(
+            name="list-storage",
+            image="docker.io/ocdr/dkube-datascience-tf-cpu:v2.0.0-3",
```
Comment (mak-454) [inline external-data/external-data.py:67]: can use some smaller image here ? say ubuntu or alpine ?

File: external-data/config.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  url: https://github.com/oneconvergence/dkube-examples.git
+  branch: pipelines
+  private: false
+datasets:
+  - name: heart-data
+    source: aws_s3
+    bucket: dkube
+    prefix: datasets/heart-data/
+    key: __aws-key__
+    secret: __aws-secret__
```
Comment (mak-454) [inline external-data/config.yaml:12]: there should remote: True

File: external-data/preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    s3_client.Bucket(config['Bucket']).download_file('datasets/heart-data/heart.csv', 'heart.csv')
+
+    ########--- Parse for parameters ---########
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--url", dest="url", default=None, type=str, help="setup URL")
+    parser.add_argument("--train_fs", dest="train_fs", required=True, type=str, help="train featureset")
+    global FLAGS
+    FLAGS, unparsed = parser.parse_known_args()
+    
+    dkubeURL = FLAGS.url
```
Comment (mak-454) [inline external-data/preprocessing.py:33]: this is not passed from config.yaml. Also not needed, as this code runs inside pod, dkube URL can be picked up from ENV

File: external-data/external-data.py (hunk 1 above)
Comment (mak-454) [reply external-data/external-data.py:67]: is there a better way to check if certain files exist or not as oppose to just "ls"

File: multimodel/multimodel-serving.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+dkube_serving_op = component_store.load_component("serving")
+storage_op = component_store.load_component("storage")
+
+
+@kfp.dsl.pipeline(
+    name="multimodel-pl", description="sample multimodel pipeline with dkube components"
+)
+def multimodel_pipeline(model):
+
+    with kfp.dsl.ExitHandler(
```
Comment (mak-454) [inline multimodel/multimodel-serving.py:18]: we may not need storage op for this testcase

File: multiple-parallel/multiple-parallel.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import kfp
+
+@kfp.dsl.pipeline(
+    name='Parallel stages pipeline',
+    description='An example pipeline to launch number of stages in parallel'
+)
+def parallel_pipeline():
+    count = 22
```
Comment (mak-454) [inline multiple-parallel/multiple-parallel.py:8]: can count be an input ?

File: storageop-testcases/config.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ 
+datasets:
+  - name: storageop-remote-data
+    random: False
+    source: s3
+    remote: True
+    endpoint: https://download.ce8i5.arvadosapi.com
+    bucket: ce8i5-4zz18-j45o88d58u7js60
+    prefix: ""
+    key: 3dnjzwsn4y8mhg03u0mdpucp2jrdcwlskvpmrd1ytdvfippou8
```
Comment (mak-454) [inline storageop-testcases/config.yaml:15]: dkube-examples is public correct.. can we not commit keys/secret here ?

File: storageop-testcases/storageop-tc.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+        storage = storage_op(
+            "export",
+            auth_token=str(token),
+            namespace="kubeflow",
+            input_volumes=input_volumes  
+        ).after(download_data)
+        train = kfp.dsl.ContainerOp(
+            name="container-op",
+            image="docker.io/ocdr/dkube-datascience-tf-cpu:v2.0.0-3",
```
Comment (mak-454) [inline storageop-testcases/storageop-tc.py:65]: use smaller image here.. alpine or ubuntu

Comment (mak-454) [reply storageop-testcases/storageop-tc.py:65]: or can we do some program here like transform the input

**Comment by mak-454** [comment]
1 - Why is config.yaml which is required for robot tests are here ? Because of which values for some of the variables like key, secret are getting exposed here.
@deepio-oc ?

2 - Storage op testcases related to Output and Intermediate volumes are missing.

3 - Metrics Op pipeline /testcase is missing

**Comment by deepio-oc** [comment]
@mak-454 these pipeline examples serves robot pipeline tests as well as users who are coming here to see different pipeline usage. 

we are keeping config.yaml so that users who are seeing the pipelines knows what arguments to pass to run these pipelines. We generally use placeholder for secrets which robot tests fill.  The credentials you mentioned, they are already public in dkubeio-examples (or dkube-examples-internal) I think.

One more reason to maintain config.yaml here is that pipeline development and the runtime arguments goes together, so should be at one place

**Comment by mak-454** [comment]
> One more reason to maintain config.yaml here is that pipeline development and the runtime arguments goes together, so should be at one place

Pipeline parameters are for this purpose correct. Also if not for robot, it is not very clear from config.yaml that those resources must be created before running this pipeline. Also it is not clear how to create them.

**Comment by deepio-oc** [comment]
I didnt understand what you meant by `Pipeline parameters are for this purpose correct`

config.yaml serves a reference for pipeline arguments.  Added documentation in top README file about config.yaml.

**Comment by deepio-oc** [comment]
@pallavi-pannu-oc before we merge these pipelines, I want to make sure each pipelines is showing a use case not covered by other pipelines.

Changing a dataset type from local to remote or to custom S3, doesn't change pipeline (or even number of pipeline arguments). 

Please list down the pipelines added and what is the unique in that pipeline which are not covered in other pipeline.

File: limits-requests/config.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+code:
+  name: dkube-examples
+  url: https://github.com/oneconvergence/dkube-examples.git
+  branch: tensorflow
+  private: false
+  
+arguments:
+  code: dkube-examples
```
Comment (deepio-oc) [inline limits-requests/config.yaml:8]: no code is required for this

File: limits-requests/limits-requests.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+dkube_training_op = component_store.load_component("training")
+
+
+@kfp.dsl.pipeline(name="set-envvars-pl", description="pl with limits and requests ")
+def limits_requests_pipeline(
+    code, run_script, limit_cpu, limit_memory, request_cpu, request_memory, token
+):
+
+    a = [
```
Comment (deepio-oc) [inline limits-requests/limits-requests.py:17]: a=>envs

**Comment by deepio-oc** [comment]
@pallavi-pannu-oc could you close this. @rahul is raising individual PRs for pipelines

**Pull request closed without merging**